src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -1098,6 +1078,70 @@ public async Task WebApplication_CallsUseRoutingAndUseEndpoints()
             Assert.Equal("One", chosenRoute);
         }
 
+        [Fact]
+        public async Task BranchingPipelineHasOwnRoutes()
+        {
+            var builder = WebApplication.CreateBuilder();
+            builder.WebHost.UseTestServer();
+            await using var app = builder.Build();
+
+            app.UseRouting();
+
+            var chosenRoute = string.Empty;
+            app.MapGet("/", () => "Hello World!").WithDisplayName("One");
+
+            app.UseEndpoints(routes =>
+            {
+                routes.MapGet("/hi", async c =>
+                {
+                    chosenRoute = c.GetEndpoint()?.DisplayName;
+                    await c.Response.WriteAsync("Hello World");
+                }).WithDisplayName("Two");
+                routes.MapGet("/heyo", () => "Heyo World").WithDisplayName("Three");
+            });
+
+            // REVIEW: this cast is strange
```

Comment:
It's not strange, most extension methods are written against IApplicationBuilder. We hide New by default on WebApplication because it's a bit weird to expose.